File: backend/src/main/kotlin/challenge/model/Order.kt
package model

import java.util.*

class Order(val customer: Customer, val address: Address) {
    private val items = mutableListOf<OrderItem>()
    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount: Double
        get() = items.sumOf { it.total }

    fun addProduct(product: Product, quantity: Int) {
        val productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception("The product has already been added. Change the amount if you want more.")

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null)
            throw Exception("The order has already been paid!")

        if (items.isEmpty())
            throw Exception("Empty order cannot be paid!")

        payment = Payment(this, method)

        // Process each item with its respective strategy
        items.forEach { item ->
            val strategy = when (item.product.type) {
                ProductType.PHYSICAL -> PhysicalProductPaymentProcessing()
                ProductType.BOOK -> BookPaymentProcessing()
                ProductType.DIGITAL -> DigitalMediaPaymentProcessing()
                ProductType.MEMBERSHIP -> MembershipPaymentProcessing()
            }
            strategy.processPayment(this)
        }

        close()
    }

    private fun close() {
        closedAt = Date()
        println("Order closed at $closedAt")
    }
}

File: backend/individual-assignment/src/models/Order.ts
import { OrderItem } from './OrderItem';
import { Customer } from './Customer';
import { Address } from './Address';
import { Payment } from './Payment';
import { Product } from './Product';
import{PaymentStrategy} from '../interfaces/PaymentStrategy';

export class Order {
    public items: OrderItem[] = [];
    public closedAt: Date | null = null;
    public payment: Payment | null = null;

    constructor(public customer: Customer, public address: Address) {}

    get totalAmount(): number {
        return this.items.reduce((sum, item) => sum + item.total, 0);
    }

    addProduct(product: Product, quantity: number): void {
        const productAlreadyAdded = this.items.some(item => item.product === product);
        if (productAlreadyAdded) {
            throw new Error("The product has already been added. Change the amount if you want more.");
        }

        this.items.push(new OrderItem(product, quantity));
    }

    pay(strategy: PaymentStrategy): void {
        if (this.payment) {
            throw new Error("The order has already been paid!");
        }

        if (this.items.length === 0) {
            throw new Error("Empty order cannot be paid!");
        }

        strategy.processPayment(this);
        this.close();
    }

    public close(): void {
        this.closedAt = new Date();
    }
}
